python/hail/methods/misc.py

```diff
@@ -358,4 +368,213 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+    ``starts`` and ``stops``, each of size equal to the number of rows.
+
+    By default, for all indices ``i``, ``[starts[i], stops[i])`` is the maximal
+    range of row indices ``j`` such that ``contig[i] == contig[j]`` and
+    ``position[i] - radius <= position[j] <= position[i] + radius``.
+
+    If the :meth:`.global_position` on `locus_expr` is not in ascending order,
+    this method will fail. Ascending order should hold for a matrix table keyed
+    by locus or variant (and the associated row table), or for a table that's
+    been ordered by `locus_expr`.
+
+    Set `coord_expr` to use a value other than position to define the windows.
+    This row-indexed numeric expression must be non-missing, non-``nan``, on the
+    same source as `locus_expr`, and non-decreasing with respect to locus
+    position for each contig; otherwise the function will fail.
+
+    The last example above uses centimorgan coordinates, so
+    ``[starts[i], stops[i])`` is the maximal range of row indices ``j`` such
+    that ``contig[i] == contig[j]`` and
+    ``cm[i] - radius <= cm[j] <= cm[i] + radius``.
+
+    Index ranges are start-inclusive and stop-exclusive. This function is
+    especially useful in conjunction with
+    :meth:`.BlockMatrix.sparsify_row_intervals`.
+
+    Parameters
+    ----------
+    locus_expr : :class:`.LocusExpression`
+        Row-indexed locus expression on a table or matrix table.
+    radius: :obj:`int`
+        Radius of window for row values.
+    coord_expr: :class:`.Float64Expression`, optional
+        Row-indexed numeric expression for the row value.
+        Must be on the same table or matrix table as `locus_expr`.
+        By default, the row value is given by the locus position.
+
+    Returns
+    -------
+    (:class:`ndarray` of :obj:`int64`, :class:`ndarray` of :obj:`int64`)
+        Tuple of start indices array and stop indices array.
+    """
+    if radius < 0:
+        raise ValueError(f'locus_windows: radius must be non-negative, found {radius}')
+
+    check_row_indexed('locus_windows', locus_expr)
+
+    if coord_expr is None:
+        global_pos = np.array(locus_expr.global_position().collect(), dtype=np.int64)
+        coord = global_pos
+        n_loci = coord.size
+    else:
+        check_row_indexed('locus_windows', coord_expr)
+        global_pos_and_coord =\
+            hl.tuple([locus_expr.global_position(), coord_expr]).collect()  # raises exception if sources differ
+        n_loci = len(global_pos_and_coord)
+
+        global_pos = np.zeros(n_loci, dtype=np.int64)
+        coord = np.zeros(n_loci, dtype=np.float64)
+        for i in range(n_loci):
+            global_pos[i] = global_pos_and_coord[i][0]
+            coord[i] = global_pos_and_coord[i][1]
+        del global_pos_and_coord
+
+    if n_loci == 0:
+        return np.zeros(shape=0, dtype=np.int64), np.zeros(shape=0, dtype=np.int64)
+
+    if not np.all(global_pos[:-1] <= global_pos[1:]):
```

Comment:
I moved the check in `locus_windows` into the for loop. I left the check in `array_windows` as separate since it's more complicated to combine and takes a few microseconds even for millions of elements.